### PR TITLE
[TEP-0104] Update Pod with Task-level Resource Requirements

### PR DESCRIPTION
**Pull request by austinzhao-go**
# Changes
/kind feature

Closes https://github.com/tektoncd/pipeline/issues/4470

The related TEP:
[TEP-0104: Task-level Resource Requirements](https://github.com/tektoncd/community/blob/main/teps/0104-tasklevel-resource-requirements.md)

The related impl PRs:
https://github.com/tektoncd/pipeline/pull/4877 - Fields Addition & Validation w/ Docs Updates
https://github.com/tektoncd/pipeline/pull/5054 - Add Validation for Step-level Resource Requirements 
https://github.com/tektoncd/pipeline/pull/5212 - Populate Task-level Resource Requirements from PipelineRun to TaskRun

This PR will include logic and tests as:
- Update the containers in the Pod with the specified task-level resource requirements

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing (included in https://github.com/tektoncd/pipeline/pull/4877)
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in (if there are no user facing changes, use release note "NONE")

# Release Notes
``` release-note
Supported Task-level Resource Requirements 
```

### CODE REVIEW

**Comment by austinzhao-go** [comment]
/hold

till https://github.com/tektoncd/pipeline/pull/5054 got merged

File: pkg/pod/pod.go
Comment (lbernick) [inline pkg/pod/pod.go:371]: in the TEP i have the resources applied only to the first container, but I think it would actually be better to split them among the containers. The reason is that limitrange default requests are split among the containers. consider this example:

2 steps
task level memory: 1Gi
limitrange default requests: 600Mi

If you split the task level resources between containers you'll have each container with 500Mi (which is what we want). If you just apply them to one container, the limitrange defaults will kick in for the container without resources and you'll have one container with 1Gi and one with 300Mi (not what we want). 

I'll update the TEP to clarify this.

File: pkg/pod/pod_test.go
Comment (austinzhao-go) [reply pkg/pod/pod_test.go:2068]: nice hint. think I need to dig a bit and try some cases to ensure it.

File: pkg/pod/pod.go
Comment (lbernick) [inline pkg/pod/pod.go:372]: I don't think you need steptemplate here. Steps are merged with step template and overrides [here](https://github.com/tektoncd/pipeline/blob/3f539724f27a6add8b64d0712e056e4479750975/pkg/pod/pod.go#L136-L149)

File: pkg/pod/pod_test.go
Comment (lbernick) [inline pkg/pod/pod_test.go:2088]: I think this is a mistake in the TEP :( 

Here, this says it's invalid: https://github.com/tektoncd/community/blob/main/teps/0104-tasklevel-resource-requirements.md#interaction-with-step-resource-requirements

However this example uses it: https://github.com/tektoncd/community/blob/main/teps/0104-tasklevel-resource-requirements.md#example-with-steptemplate

Would you be comfortable taking on a PR to the TEP fixing this mistake?

**Review by lbernick — COMMENTED** [review]
I know this needs to be rebased onto #5054, but could you keep the scope of this PR to just the changes required to apply the computeResources of a taskrun to a pod? i.e. validation and passing spec from pipelinerun -> taskrun should be covered in #5054. 
Also, I think a better place for some of this logic might be pkg/internal/computeresources.

**Comment by austinzhao-go** [comment]
> 

yap. let me keep validation related in https://github.com/tektoncd/pipeline/pull/5054.

File: pkg/pod/pod.go
Comment (austinzhao-go) [reply pkg/pod/pod.go:372]: I see. thanks this context!
let me clean it up.

File: pkg/pod/pod_test.go
Comment (austinzhao-go) [reply pkg/pod/pod_test.go:2088]: got this. think we still will be good for overwriting resource requirements from StepTemplate while validating/avoiding ones from Step? 

yap. I can do this. (will raise today)

Comment (lbernick) [reply pkg/pod/pod_test.go:2088]: Awesome thank you! I think if you can't specify compute resources on a step and a taskrun, you also shouldn't be able to do it on a steptemplate and a taskrun.

Comment (austinzhao-go) [reply pkg/pod/pod_test.go:2088]: right! I got your whole point now - as stepTemplate is for step.. 
thanks Lee, will mark it invalid - raise TEP update PR - then update code

Comment (austinzhao-go) [reply pkg/pod/pod_test.go]: raised update PR https://github.com/tektoncd/community/pull/763

re-arranged validation in https://github.com/tektoncd/pipeline/pull/5054
and removed the StepTemplate case for pod_test.go

File: pkg/internal/computeresources/transformer.go
Comment (austinzhao-go) [inline pkg/internal/computeresources/transformer.go:118]: think some discussion will come to this line:

my current understanding as 
- if not limitranges, just stay as it is

- if currentRequest < min (of limitrange), keep the min
- if over the limit, set as the limit

for limits
- similarly, but keep the min of currentLimit and limit from limitrange

File: pkg/internal/computeresources/transformer_test.go
Comment (austinzhao-go) [inline pkg/internal/computeresources/transformer_test.go:316]: these updates on previous testcases are to align with updated logics on "setting requests/limits with limitranges"

File: pkg/internal/computeresources/tasklevel/tasklevel_test.go
Comment (lbernick) [inline pkg/internal/computeresources/tasklevel/tasklevel_test.go:128]: StepTemplate isn't used in this test at all, so this test isn't testing what we want it to. At the point where `ApplyTaskLevelComputeResources` is called, the step template resource requests have already been applied to the input containers [here](https://github.com/tektoncd/pipeline/blob/e8ee8798a6758bd4bfdd41bb23ec63eff1fb7aec/pkg/pod/pod.go#L138). There are two options here:
1. add a test case where the input steps already have resources on them, and make sure they don't make their way to the output. I don't love this solution because it relies on the behavior I linked to.
2. Create an overall pod test with steptemplate resources and task level resources, and make sure the pod created has the right compute resources. This would be a more thorough test but harder to work with.

I'll let you use your best judgment here but I'll also try to figure out if there's any refactoring I can try here.

File: pkg/internal/computeresources/tasklevel/tasklevel_test.go
Comment (lbernick) [inline pkg/internal/computeresources/tasklevel/tasklevel_test.go:167]: Same comment as above; sidecars are ignored in `ApplyTaskLevelComputeResources` so they're not actually being tested here.

File: pkg/internal/computeresources/transformer_test.go
Comment (lbernick) [inline pkg/internal/computeresources/transformer_test.go:392]: Was this a bug? Could you make this change in a separate PR?

File: pkg/internal/computeresources/tasklevel/tasklevel_test.go
Comment (austinzhao-go) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go:81]: thanks for more thought on this point. 

I tend to take the "request as the existing limit when only limit is configured" way
- think will align with current k8s behaviours - so ease user's understanding and usages
- and "0 / smallest as possible" request could lead to a "resource hungry" case in the runtime, like the CPU requests acting as weights to acquire CPU time as https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#how-pods-with-resource-limits-are-run

File: pkg/internal/computeresources/tasklevel/tasklevel_test.go
Comment (austinzhao-go) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go:167]: ya. this testcase actually came from one of our previous threads
https://github.com/tektoncd/pipeline/pull/4877#discussion_r899154954

File: pkg/internal/computeresources/transformer_test.go
Comment (austinzhao-go) [reply pkg/internal/computeresources/transformer_test.go:392]: sure. and saw your raised refact PR.

think good for me to keep "updating pod" logic in this PR while leaving the transformer and limitrange changes.

File: pkg/internal/computeresources/tasklevel/tasklevel_test.go
Comment (austinzhao-go) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go:128]: thanks this context. think good for me mark as a follow-up action.
and will remove this case at this moment.

File: pkg/internal/computeresources/transformer_test.go
Comment (austinzhao-go) [reply pkg/internal/computeresources/transformer_test.go:392]: a follow-up PR I will apply on "setting the requests and limits" after this refact PR was merged.

Comment (austinzhao-go) [reply pkg/internal/computeresources/transformer_test.go]: raised here - https://github.com/tektoncd/pipeline/pull/5269

File: pkg/internal/computeresources/tasklevel/tasklevel_test.go
Comment (austinzhao-go) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go]: closed - moved this testcase under pod_test.go

**Comment by austinzhao-go** [comment]
/unhold

as depended PRs were merged/closed

File: pkg/pod/pod_test.go
Comment (lbernick) [inline pkg/pod/pod_test.go:2009]: this looks great! could you also add a test for where the steps have resource requirements?

File: pkg/pod/pod_test.go
Comment (lbernick) [inline pkg/pod/pod_test.go:2138]: 👍 could you please also check the container name? This will help us test that the sidecar is getting the right resource requirements, as opposed to the step containers

File: pkg/internal/computeresources/tasklevel/tasklevel_test.go
Comment (lbernick) [inline pkg/internal/computeresources/tasklevel/tasklevel_test.go:34]: Since the function being tested accepts a slice of steps, I think that's what should be in the test case. (I don't think it makes sense to use a TaskSpec that has other fields including sidecars, because this function doesn't accept task specs or sidecars in its input.

File: pkg/internal/computeresources/tasklevel/tasklevel.go
Comment (lbernick) [inline pkg/internal/computeresources/tasklevel/tasklevel.go:25]:
```suggestion
// ApplyTaskLevelComputeResources applies the task-level compute resource requirements to each Step
```

File: pkg/internal/computeresources/tasklevel/tasklevel_test.go
Comment (lbernick) [inline pkg/internal/computeresources/tasklevel/tasklevel_test.go:31]:
```suggestion
func TestApplyTaskLevelResourceRequirements(t *testing.T) {
```

File: pkg/internal/computeresources/tasklevel/tasklevel_test.go
Comment (lbernick) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go:81]: Good point that this could lead to a pod being more likely to be evicted. I'll update the TEP. in the meantime, I think you can go ahead with implementing this strategy.

**Comment by austinzhao-go** [comment]
/hold

till https://github.com/tektoncd/pipeline/pull/5269 got merged

File: pkg/internal/computeresources/tasklevel/tasklevel_test.go
Comment (lbernick) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go:81]: https://github.com/tektoncd/community/pull/776

File: pkg/internal/computeresources/tasklevel/tasklevel_test.go
Comment (austinzhao-go) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go:34]: the tested function here we mean `builder.Build()` right? which needs a TaksRun and a TaskSepc

Comment (austinzhao-go) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go:34]: the `Build` function is the one we mentioned in last week to test as it covers logic related to steps, stepTemplate, sidecar to form the container configs to create the Pod. 
`verifyTaskLevelComputeResources` is the help func which we need to verify if the result pod matches our expectation.

Comment (lbernick) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go:34]: It looks like this function is testing ApplyTaskLevelResourceRequirements not builder.Build

File: pkg/internal/computeresources/tasklevel/tasklevel_test.go
Comment (austinzhao-go) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go:31]: as above thread - think the tested func is `Build`

File: pkg/internal/computeresources/tasklevel/tasklevel_test.go
Comment (lbernick) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go:34]: I think your approach in pod_test.go (where you're using the TaskSpec and calling builder.Build) is correct!

Comment (lbernick) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go:34]: Unresolving this conversation, because while the tests in pod_test.go LGTM (the ones testing builder.Build), this test (in tasklevel_test.go) is testing the function `tasklevel.ApplyTaskLevelComputeResources`, which only takes steps and compute resources. Therefore, I think the test cases in this file should use steps and compute resources as their inputs, not task specs with other fields in addition.

File: pkg/internal/computeresources/tasklevel/tasklevel_test.go
Comment (lbernick) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go:81]: Unresolving since I don't think this comment was addressed; just to be clear I'm suggesting implementing what's proposed in the TEP PR.

Comment (austinzhao-go) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go:81]: thought this will be handled in limitrange (wrongly). but a case will be if there is no limitrange.
let me update

File: pkg/internal/computeresources/tasklevel/tasklevel_test.go
Comment (austinzhao-go) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go:34]: missed these 2 files. sry for this. 
thanks the double check - let me update!

File: pkg/internal/computeresources/tasklevel/tasklevel_test.go
Comment (austinzhao-go) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go]: added the required logic and updated testcases.

File: pkg/internal/computeresources/tasklevel/tasklevel_test.go
Comment (austinzhao-go) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go]: cleaned testcases struct.

File: pkg/internal/computeresources/tasklevel/tasklevel_test.go
Comment (lbernick) [inline pkg/internal/computeresources/tasklevel/tasklevel_test.go:71]: Since the limits are "500m", we also want to set the total requests to "500m", meaning that each container should have requests of "250m".

Comment (austinzhao-go) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go:71]: I see. this (limits) will be applied as the total task-level allowed requests.
let me correct.

Comment (austinzhao-go) [reply pkg/internal/computeresources/tasklevel/tasklevel_test.go]: corrected by "spreading" the limits when no requests.

File: pkg/internal/computeresources/tasklevel/tasklevel.go
Comment (lbernick) [inline pkg/internal/computeresources/tasklevel/tasklevel.go:38]: nit: no need to compute this on every iteration

**Review by lbernick — APPROVED** [review]
This looks great!
A few notes:
- once #5269 is merged I think we should add some unit tests to pod_test.go that use the limitrange transformer
- We will need to wait until https://github.com/tektoncd/community/pull/776 is merged as well
- Please update the docs to clarify the case where a user sets limits but no requests

File: pkg/internal/computeresources/tasklevel/tasklevel.go
Comment (austinzhao-go) [reply pkg/internal/computeresources/tasklevel/tasklevel.go]: replaced with a var - thanks this!

File: docs/compute-resources.md
Comment (lbernick) [inline docs/compute-resources.md:88]: What we specifically want to clarify here is that if no requests are set, the *task-level* requests will be set to the *task-level* limits

File: docs/compute-resources.md
Comment (lbernick) [inline docs/compute-resources.md:66]: Also, please update line 48 to clarify that this feature is implemented!

File: docs/compute-resources.md
Comment (austinzhao-go) [reply docs/compute-resources.md:88]: right. misunderstood your original point. - quick update

File: docs/compute-resources.md
Comment (austinzhao-go) [reply docs/compute-resources.md:66]: yap.

File: docs/compute-resources.md
Comment (lbernick) [inline docs/compute-resources.md:48]: sorry I should have been more clear-- this line can just be removed. (Users can safely assume everything in our docs is implemented unless specified otherwise.

**Pull request merged**
